drivers/samsung/displays/mdc_protocol.cr

```diff
@@ -0,0 +1,376 @@
+module Samsung; end
+
+require "placeos-driver/interface/powerable"
+require "placeos-driver/interface/muteable"
+require "placeos-driver/interface/switchable"
+
+class Samsung::Displays::MDCProtocol < PlaceOS::Driver
+  include Interface::Powerable
+  include Interface::Muteable
+
+  INDICATOR = 0xAA_u8
+
+  enum Inputs
+    Vga           = 0x14 # pc in manual
+    Dvi           = 0x18
+    Dvi_Video     = 0x1F
+    Hdmi          = 0x21
+    Hdmi_Pc       = 0x22
+    Hdmi2         = 0x23
+    Hdmi2_Pc      = 0x24
+    Hdmi3         = 0x31
+    Hdmi3_Pc      = 0x32
+    Hdmi4         = 0x33
+    Hdmi4_Pc      = 0x34
+    Display_Port  = 0x25
+    Dtv           = 0x40
+    Media         = 0x60
+    Widi          = 0x61
+    Magic_Info    = 0x20
+    Whiteboard    = 0x64
+  end
+  include PlaceOS::Driver::Interface::InputSelection(Inputs)
+
+  # Discovery Information
+  tcp_port 1515
+  descriptive_name "Samsung MD, DM & QM Series LCD"
+  generic_name :Display
+
+  # Markdown description
+  description <<-DESC
+  For DM displays configure the following 1:
+
+  1. Network Standby = ON
+  2. Set Auto Standby = OFF
+  3. Set Eco Solution, Auto Off = OFF
+
+  Hard Power off displays each night and hard power ON in the morning.
+  DESC
+
+  default_settings({
+    display_id: 0,
+    rs232_control: false
+  })
+
+  @id : Int32 = 0
+  @rs232 : Bool = false
+  @blank : Input?
+  @previous_volume : Int32 = 50
+  # Meta data for inquiring interfaces
+  @input_stable : Bool = true
+  @input_target : Input = Input::Hdmi
+  @power_stable : Bool = true
+
+  def on_load
+    transport.tokenizer = Tokenizer.new do |io|
+      bytes = io.peek
+      # Ensure message indicator is well-formed
+      disconnect unless bytes.first == INDICATOR
+      logger.debug { "Received: #{bytes}" }
+      # [header, command, id, data.size, [data], checksum]
+      # return 0 if the message is incomplete
+      bytes.size < 4 ? 0 : bytes[3].to_i + 5
+    end
+
+    on_update
+  end
+
+  def on_update
+    @id = setting(Int32, :display_id)
+    @rs232 = setting(Bool, :rs232_control)
+    if blanking_input = setting?(String, :blanking_input)
+      @blank = Input.parse?(blanking_input)
+    end
+  end
+
+  def connected
+    do_device_config unless self[:hard_off]?
+
+    schedule.every(30.seconds, true) do
+      do_poll
+    end
+  end
+
+  def disconnected
+    self[:power] = false unless @rs232
+    schedule.clear
+  end
+
+  # As true power off disconnects the server we only want to
+  # power off the panel. This doesn't work in video walls
+  # so if a nominal blank input is
+  # TODO: check type for broadcast is correct
```

Comment:
This TODO should be resolved before merge.

Reply:
I've removed broadcast as that was from the old ruby implementation. Can add it back later if it's needed. https://github.com/PlaceOS/drivers/pull/19/commits/630d1213b538f2702af767969d2bd9a33d6597d7